File: api/app/routers/publications.py
import re

from fastapi import APIRouter, HTTPException, Query

from ..database import db

router = APIRouter(prefix="/publications", tags=["Publicações"])


@router.get("")
async def list_publications(channel: str | None = None, limit: int = Query(24, ge=1, le=100), offset: int = Query(0, ge=0)):
    params = {"select": "id,campaign_id,channel,title,slug,public_path,payload,seo,published_at", "status": "eq.published", "is_public": "eq.true", "order": "published_at.desc", "limit": str(limit), "offset": str(offset)}
    if channel:
        params["channel"] = f"eq.{channel}"
    return await db.request("GET", "communication_items", params=params)


@router.get("/{slug_or_id}")
async def get_publication(slug_or_id: str, channel: str | None = None):
    if not re.fullmatch(r"[a-zA-Z0-9-]{1,180}", slug_or_id):
        raise HTTPException(status_code=404, detail="Publicação não encontrada.")
    params = {"select": "id,campaign_id,channel,title,slug,public_path,payload,seo,published_at", "status": "eq.published", "is_public": "eq.true", "limit": "1"}
    params["id" if re.fullmatch(r"[0-9a-fA-F-]{36}", slug_or_id) else "slug"] = f"eq.{slug_or_id}"
    if channel:
        params["channel"] = f"eq.{channel}"
    rows = await db.request("GET", "communication_items", params=params)
    if not rows:
        raise HTTPException(status_code=404, detail="Publicação não encontrada.")
    return rows[0]

File: src/features/omnipublish/publications.ts
import 'server-only';

import { createAdminClient } from '@/lib/supabase/admin';
import type { CommunicationChannel } from '@/features/admin/omnipublish/types';

export async function loadPublicCatalog(channel: CommunicationChannel) {
  const supabase = createAdminClient();
  const [{ data: catalog }, { data: items }] = await Promise.all([
    supabase.from('communication_catalogs').select('*').eq('channel', channel).eq('is_public', true).maybeSingle(),
    supabase.from('communication_items').select('id, title, slug, public_path, payload, seo, published_at, communication_campaigns!inner(summary, campaign_tags, status)').eq('channel', channel).eq('status', 'published').eq('is_public', true).eq('communication_campaigns.status', 'published').order('published_at', { ascending: false }),
  ]);
  return { catalog, items: items ?? [] };
}

export async function loadPublicPublication(channel: CommunicationChannel, slugOrId: string) {
  const supabase = createAdminClient();
  let query = supabase.from('communication_items').select('id, title, slug, public_path, payload, seo, published_at, communication_campaigns!inner(summary, campaign_tags, status)').eq('channel', channel).eq('status', 'published').eq('is_public', true).eq('communication_campaigns.status', 'published');
  query = /^[0-9a-f-]{36}$/i.test(slugOrId) ? query.eq('id', slugOrId) : query.eq('slug', slugOrId);
  const { data } = await query.maybeSingle();
  return data;
}
